Initial Implementation of 'Find Meeting Times' Algorithm

zghera opened the pull request:
Implement initial algorithm that finds all potential meeting time ranges with a set of participants, duration of meeting, and set of existing events.

This algorithm implemented the query method in FindMeetingQuery class in order to pass all of the pre-written tests in FindMeetingQueryTest.java. The algorithm's asymptotic time complexity is bottle-necked by Collections.sort() which is worst case O(n log n), where n is the number of events. This is only the case if we assume that the number of attendees in each event is limited to a fairly small number. If the average number of attendees within each event cannot be assumed to be small, then the worst case time complexity would be O(n*m) where m is the number of attendees for a single event.

The general steps in the algorithm:
- Remove any events where it does not contains at least one participant that is included in the meeting request.
- Sort the remaining list of events based on start time.
- Create a start and end of day reference points (create an int for start time and add end time to the event list) to avoid having to add additional conditional statements.
- Loop through each event and see if the amount of time between the current event start time and the earlier event end time is greater than or equal to the requested meeting duration.

As mentioned above the initial algorithm was tested with unit tests that were pre-written to test the query method. In a future PR, more tests will be added when implementing optional attendees.

zghera committed:
Initial implementation of 'find meeting times' algorithm. Passes all pre-written tests.

beiebi0 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,72 @@
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
+  private boolean eventParticipantInMeeting(Set<String> eventAttendees, 
+                                              Collection<String> meetingAttendees) {
+    for (String eventAttendee : eventAttendees) {
+      if (meetingAttendees.contains(eventAttendee)) {
+        return true;
+      }
+    }
+    return false;
+  }
```
consider to use set.retainAll() function to do this intersection

zghera replied:
This is a pretty cool function! Current implementation uses this now.

beiebi0 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,72 @@
+    // Remove any events where there are no participants in event that are in meeting request.
+    for(Iterator<Event> it = eventList.iterator(); it.hasNext();) {
+      Event cur = it.next();
+      if (eventParticipantInMeeting(cur.getAttendees(), request.getAttendees()) == false) {
```
```suggestion
      if (!eventParticipantInMeeting(cur.getAttendees(), request.getAttendees())) {
```

beiebi0 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,72 @@
+    List<Event> eventList = new ArrayList<>(events);
+
+    // Remove any events where there are no participants in event that are in meeting request.
+    for(Iterator<Event> it = eventList.iterator(); it.hasNext();) {
+      Event cur = it.next();
+      if (eventParticipantInMeeting(cur.getAttendees(), request.getAttendees()) == false) {
```
consider to use filter() from Stream https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html

zghera replied:
Oh wow this is very nice, thanks for the tip! Done.

beiebi0 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,72 @@
+        it.remove();
+      }
+    }
+    Collections.sort(eventList, new Comparator<Event>() {
+      @Override
+      public int compare(Event a, Event b) {
+        return Long.compare(a.getWhen().start(), b.getWhen().start());
+      }
+    });
```
can call sorted() from Stream as well.

zghera replied:
Done.

beiebi0 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,72 @@
+      Event curEvent = eventList.get(i);
+      int startOfCurEvent = curEvent.getWhen().start();
+
+      boolean endTimeIsInclusive = false;
+      if ("EOD".equals(curEvent.getTitle())) {
+        endTimeIsInclusive = true;
+      }
```
can be 
```suggestion
      boolean endTimeIsInclusive = "EOD".equals(curEvent.getTitle());
```

zghera replied:
Yeah this is a lot cleaner. Done.

zghera committed:
Added documentation and broke query into more private helpers.

zghera committed:
Use retainAll to determine whether or not event participants intersect with meeting participants.

zghera committed:
Refactor filtering of events with no meeting participants.

Use a stream of event objects with filter rather than manually doing this with an iterator.

zghera committed:
Call sorted from stream rather than using Collections.sort.

zghera committed:
Remove un-needed comments.

zghera committed:
Simplify bool to determine if the cur event is EOD.

keiffer01 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,106 @@
+  private boolean eventParticipantInMeeting(
+      Set<String> eventAttendeesCopy, Collection<String> meetingAttendees) {
+    eventAttendeesCopy.retainAll(meetingAttendees);
+    if (eventAttendeesCopy.isEmpty()) {
+      return false;
+    }
+    return true;
```
```suggestion
    return !eventAttendeesCopy.isEmpty()
```

zghera replied:
Much cleaner, thanks!

keiffer01 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,106 @@
+   * @param request The meeting request that contains the requirements for potential meetings.
+   * @return A Collection of the feasible meeting {@code TimeRange}s.
+   */
+  private Collection<TimeRange> getMeettingTimes(List<Event> eventList, MeetingRequest request) {
```
Typo?
```suggestion
  private Collection<TimeRange> getMeetingTimes(List<Event> eventList, MeetingRequest request) {
```

zghera replied:
Ahh yes, spelling is hard. Thanks for the catch.

keiffer01 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,106 @@
+    int endOfEarlierEvent = TimeRange.START_OF_DAY;
+
+    Collection<TimeRange> openMeetingTimes = new ArrayList<>();
+    for (int i = 0; i < eventList.size(); i++) {
+      Event curEvent = eventList.get(i);
```
Could use for-each loop instead?
```suggestion
    for (Event curEvent : eventList) {
```

zghera replied:
Ahh yes. Initially I started from i = 1 so that is why I couldn't use for-each but now I can so thanks for the reminder!

keiffer01 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,106 @@
+                })
+            .collect(Collectors.toList());
+
+    return getMeettingTimes(eventList, request);
```
Typo again?
```suggestion
    return getMeetingTimes(eventList, request);
```

zghera replied:
Same as above.

keiffer01 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,106 @@
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> eventList =
+        events.stream()
+            .filter(
+                event -> eventParticipantInMeeting(
+                    new HashSet<String>(event.getAttendees()), request.getAttendees()))
+            .sorted(
+                new Comparator<Event>() {
+                  @Override
+                  public int compare(Event a, Event b) {
+                    return Long.compare(a.getWhen().start(), b.getWhen().start());
+                  }
+                })
+            .collect(Collectors.toList());
```
This method of solving the problem is very functional programming style, pretty cool!

zghera replied:
Thanks! I had never heard about stream() until Longjun told me about it but I agree that it is pretty sick.

zghera committed:
Merge branch 'find-meetings-initial-alg' of https://github.com/zghera/STEP-Portfolio into find-meetings-initial-alg

zghera committed:
Increase code style from PR comments.

Converted for to for-each loop. Condensed logic of return statement in eventParticipantInMeeting. Fix spelling.

zghera committed:
Run java formatter.

zghera committed:
Merge branch 'find-meetings-initial-alg' of github.com:zghera/STEP-Portfolio into find-meetings-initial-alg

HiramSilvey commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,100 @@
+   */
+  private Collection<TimeRange> getMeetingTimes(List<Event> eventList, MeetingRequest request) {
+    eventList.add(new Event(
+        "EOD", TimeRange.fromStartDuration(TimeRange.END_OF_DAY, 0), request.getAttendees()));
```
Since EOD is used multiple times, it should probably be separated out into a variable.

zghera replied:
Done.

HiramSilvey commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,100 @@
+
+      // Only move reference points of end of last event if the cur event end point is later
+      // than the end of the event with the latest end point so far.
+      int endOfCurEvent = curEvent.getWhen().end();
+      if (endOfCurEvent >= endOfEarlierEvent) {
+        endOfEarlierEvent = endOfCurEvent;
```
This could be rewritten as:
endOfEarlierEvent = max(endOfEarlierEvent, curEvent.getWhen().end());

If these variables are ints or longs, then the built-in max function should just work.

zghera replied:
Ahh this is much cleaner. Done.

zghera committed:
Create variable for EOD event title.

zghera committed:
Condense logic for moving 'endOfEarlierEvent' variable.
